### PR TITLE
Fix add in Week and MonthView, Add JUnit tests

### PR DESCRIPTION
**Pull request by Mahesh1772**

### CODE REVIEW

File: src/main/java/commandparser/CommandHandler.java [hunk 1]
```diff
@@ -128,11 +128,11 @@ private void handleUpdateCommand(String[] parts) {
     /**
      * Handles the "add" command to add new tasks.
      *
-     * @param parts the array of command parts
+     * @param parts the array of command parts.
      */
     private void handleAddCommand(String[] parts) {
         try {
-            StringParser.validateAddCommand(parts);
+            StringParser.validateAddCommand(parts, weekView, monthView, inMonthView);
```
Comment (Aak242) [inline src/main/java/commandparser/CommandHandler.java:135]: Thanks for updating the add command to fit the latest parameters

File: src/main/java/commandparser/StringParser.java [hunk 2]
```diff
@@ -51,13 +54,13 @@ public static String parseTaskDescription(String descriptionString) {
...
-     * @return the parsed priority level string
-     * @throws TaskManagerException if the priority level string is not valid
+     * @param priorityString the string representing the priority level.
+     * @return the parsed priority level string.
+     * @throws TaskManagerException if the priority level string is not valid.
      */
     public static String parsePriorityLevel(String priorityString) throws TaskManagerException {
         String trimmedPriority = priorityString.trim().toUpperCase();
-        if (trimmedPriority.equals("HIGH") || trimmedPriority.equals("MEDIUM") || trimmedPriority.equals("LOW")) {
+        if (trimmedPriority.equals("H") || trimmedPriority.equals("M") || trimmedPriority.equals("L")) {
```
Comment (Aak242) [inline src/main/java/commandparser/StringParser.java:63]: Great job on fixing the issue with the priority check!

File: src/main/java/commandparser/StringParser.java [hunk 3]
```diff
@@ -67,21 +70,39 @@ public static String parsePriorityLevel(String priorityString) throws TaskManage
...
+            }
+        } else {
+            LocalDate startOfWeek = weekView.getStartOfWeek();
+            LocalDate endOfWeek = startOfWeek.plusDays(6);
+            LocalDate dateForDay = weekView.getDateForDay(day);
+            if (dateForDay.isBefore(startOfWeek) || dateForDay.isAfter(endOfWeek)) {
+                throw new TaskManagerException("Invalid day for week view. Please enter a day between " +
+                        startOfWeek.getDayOfMonth() + " and " + endOfWeek.getDayOfMonth() + ".");
+            }
+        }
```
Comment (Aak242) [inline src/main/java/commandparser/StringParser.java:98]: Adding unique checks for the month and week view makes our code more robust, good job!

File: src/main/java/data/TaskManager.java [hunk 4]
```diff
@@ -409,33 +408,27 @@ public void markManager(WeekView weekView, MonthView monthView, boolean inMonthV
...
+            int dayInt) throws TaskManagerException {
         LocalDate date;
         if (inMonthView) {
-            date = monthView.getStartOfMonth().plusDays(dayInt);
+            LocalDate startOfMonth = monthView.getStartOfMonth();
+            LocalDate endOfMonth = startOfMonth.withDayOfMonth(startOfMonth.lengthOfMonth());
+            if (dayInt < 1 || dayInt > endOfMonth.getDayOfMonth()) {
+                throw new TaskManagerException("Invalid day for month view. Please enter a day between 1 and " +
+                        endOfMonth.getDayOfMonth() + ".");
+            }
```
Comment (Aak242) [inline src/main/java/data/TaskManager.java:420]: Good that more exception handling has been added to manage cases where the user enters a day that does not exist within the current month

File: src/test/java/commandparser/StringParserTest.java [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+
+import java.time.LocalDate;
+import java.time.format.DateTimeFormatter;
+
+import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+
+class StringParserTest {
```
Comment (Aak242) [inline src/test/java/commandparser/StringParserTest.java:16]: Thanks for adding comprehensive testing to the newly implemented StringParser class!

File: src/test/java/commandparser/StringParserTest.java [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+        assertDoesNotThrow(() -> {
+            LocalDate parsedDate = StringParser.parseDate(validDateString);
+            assertEquals(expectedDate, parsedDate);
+        });
+    }
+
+    @Test
+    void parseDateInvalidFormat() {
+        String invalidDateString = "03-15-2023";
+        assertThrows(TaskManagerException.class, () -> StringParser.parseDate(invalidDateString));
```
Comment (Aak242) [inline src/test/java/commandparser/StringParserTest.java:31]: Great that both positive and negative test cases are considered for most features

**Review by Aak242 — COMMENTED** [review]
More comprehensive testing, exception handling and formatting has been performed on the StringParser and WeekView sections. LGTM!

File: src/main/java/data/TaskManager.java [hunk 7]
```diff
@@ -435,12 +432,20 @@ public void markManager(WeekView weekView, MonthView monthView, boolean inMonthV
...
-                                                   boolean inMonthView, int dayInt) throws TaskManagerException {
-        LocalDate date;
-        if (inMonthView) {
-            date = monthView.getStartOfMonth().plusDays(dayInt);
-            checkIfDateInCurrentMonth(date);
+            boolean inMonthView, int dayInt) throws TaskManagerException {
+      
+        if (dayInt < 1 || dayInt > 31) {
+            throw new TaskManagerException("Invalid day number. Day must be between 1 and 31.");
+        }
```
Comment (clarencepohh) [inline src/main/java/data/TaskManager.java:439]: Good catch on invalid dates given via input!

File: src/test/java/commandparser/StringParserTest.java
Comment (clarencepohh) [inline src/test/java/commandparser/StringParserTest.java:1]: Lots of test cases!

**Review by clarencepohh — COMMENTED** [review]
Good job on the tests as well as JavaDocs. Thanks for fixing the last few PE-D bugs!

File: src/main/java/data/TaskManager.java [hunk 8]
```diff
@@ -435,12 +432,20 @@ public void markManager(WeekView weekView, MonthView monthView, boolean inMonthV
...
 
+        if (inMonthView) {
```
Comment (kyhjonathan) [inline src/main/java/data/TaskManager.java:441]: Good job on the exception handlings!

File: src/main/java/time/View.java [hunk 9]
```diff
@@ -53,4 +55,22 @@ public LocalDate getStartOfView() {
...
     }
+    /**
+     * Finds the date corresponding to a given day number within the same month
+     * and year as the start of the view.
+     *
+     * @param dayInt the day number to find the date for.
+     * @return the corresponding date.
+     * @throws TaskManagerException if the day number is invalid.
+     */
+    public LocalDate findDateFromNumber(int dayInt) throws TaskManagerException {
```
Comment (kyhjonathan) [inline src/main/java/time/View.java:66]: Nice exception handling!

File: src/test/java/ui/MonthViewTest.java [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+class MonthViewTest {
+
+    private MonthView monthView;
+    private DateTimeFormatter dateFormatter;
+
+    @BeforeEach
```
Comment (kyhjonathan) [inline src/test/java/ui/MonthViewTest.java:20]: Good job on the extensive testing of monthview!

File: src/main/java/data/TaskManager.java (hunk 7 above)
Comment (NicholasTanYY) [reply src/main/java/data/TaskManager.java:439]: Could we check that this resolves the issue for months  with less than 31 days?

File: src/test/java/commandparser/StringParserTest.java [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+        String invalidIndexString = "five";
+        assertThrows(TaskManagerException.class, () -> StringParser.parseTaskIndex(invalidIndexString));
+    }
+
+    @Test
+    void parseTaskDescriptionValidDescription() {
+        String description = "  This is a task description.  ";
+        String expectedDescription = "This is a task description.";
+        assertEquals(expectedDescription, StringParser.parseTaskDescription(description));
+    }
```
Comment (NicholasTanYY) [inline src/test/java/commandparser/StringParserTest.java:55]: Great job creating so many test cases!

File: src/test/java/ui/MonthViewTest.java [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+    }
+
+    @Test
+    void testMonthViewWithInvalidDate() {
+        assertThrows(DateTimeException.class, () -> {
+            LocalDate invalidDate = LocalDate.of(2023, 2, 30); // Invalid date
+            new MonthView(invalidDate, dateFormatter);
+        });
+    }
+}
```
Comment (NicholasTanYY) [inline src/test/java/ui/MonthViewTest.java:101]: Thank you for creating these test cases. The Month View is bound to be flawless now!

**Review by NicholasTanYY — COMMENTED** [review]
Thanks for adding many test cases. This looks ready to be merged!

**Pull request merged**